File: rust/examples/recode.rs
//! Canonical round-trip: decode framed Gen2 from stdin, re-encode RAW canonical bytes to stdout.
//! Used by the cross-language IDENTITY gate: encode(decode(wire)) must be byte-identical across
//! Go/Rust/Python/TS. JSON-projection-independent (unlike decode_cli), so it tests identity, not display.
//!
//! With `--strict` (or `STRICT=1` in the environment) the decode runs in STRICT mode
//! (SPEC-v1 §5.3): well-formed-but-non-canonical input is REJECTED (decode error -> non-zero
//! exit, empty stdout) instead of being silently accepted. Canonical input round-trips
//! identically in both modes.
//!
//! With `--addr` the tool instead prints the CONTENT ADDRESS (SPEC-v1 §3) of the canonical
//! bytes: the multihash SHA-256 (`0x12 0x20` + 32 digest bytes) of EXACTLY the canonical bytes
//! that plain `recode` would print, rendered as a single line of lowercase hex (68 chars).
//! `--addr` and `--strict` may combine.
//! With `--tensor-spans` the tool instead acts as a Phase 2 zero-copy TENSOR LOCATOR: it decodes
//! one canonical Cowrie message from stdin and prints ONE LINE PER TENSOR in document order (the
//! order tensors appear during decode), each line `<data_offset> <data_len> <dtype> <shape>`
//! (space-separated), where `<shape>` is the comma-joined shape dims (e.g. `2,3`) or `-` for a
//! rank-0 scalar. `data_offset` is the ABSOLUTE byte offset (from byte 0 of the input) of the
//! tensor's contiguous little-endian data run; `data_len` its byte length. No tensors => no
//! output. Offsets are captured during decode (not guessed), mirroring `cowrie_ref.tensor_spans`.
use cowrie_rs::gen2::{
    address_of_bytes, decode_file, decode_framed, decode_framed_strict, encode, encode_file,
    merkle_root, schema_fingerprint64, tensor_spans, to_hex,
};
use std::io::{Read, Write};

fn main() {
    let strict = std::env::args().skip(1).any(|a| a == "--strict")
        || std::env::var("STRICT").map(|v| v == "1").unwrap_or(false);
    let addr = std::env::args().skip(1).any(|a| a == "--addr");
    let file_id = std::env::args().skip(1).any(|a| a == "--file-id");
    let file_recode = std::env::args().skip(1).any(|a| a == "--file-recode");
    let tensor_spans_mode = std::env::args().skip(1).any(|a| a == "--tensor-spans");
    let fingerprint_mode = std::env::args().skip(1).any(|a| a == "--fingerprint");
    let dataset_root_mode = std::env::args().skip(1).any(|a| a == "--dataset-root");

    let mut data = Vec::new();
    std::io::stdin().read_to_end(&mut data).expect("failed to read stdin");

    // --- Dataset/stream layer Merkle root (SPEC-v1 §7 dataset_root) ---
    // Read from stdin: uvarint(n) then n repetitions of (uvarint(len) || blob_bytes).
    // Compute the §7 merkle_root over that ORDERED list of blobs (the shard file-roots) and
    // print the 34-byte multihash as 68 lowercase hex chars on one line. Reuses gen2::merkle_root.
    if dataset_root_mode {
        let mut pos = 0usize;
        let read_uvarint = |data: &[u8], p: &mut usize| -> u64 {
            let mut shift = 0u32;
            let mut val = 0u64;
            loop {
                let b = data[*p];
                *p += 1;
                val |= ((b & 0x7f) as u64) << shift;
                if b & 0x80 == 0 {
                    break;
                }
                shift += 7;
            }
            val
        };
        let n = read_uvarint(&data, &mut pos);
        let mut blobs: Vec<Vec<u8>> = Vec::with_capacity(n as usize);
        for _ in 0..n {
            let len = read_uvarint(&data, &mut pos) as usize;
            blobs.push(data[pos..pos + len].to_vec());
            pos += len;
        }
        println!("{}", to_hex(&merkle_root(&blobs)));
        return;
    }

    // --- Phase 2 zero-copy tensor locator: one line per tensor, document order ---
    if tensor_spans_mode {
        match tensor_spans(&data) {
            Ok(spans) => {
                let mut out = String::new();
                for s in &spans {
                    let shape = if s.shape.is_empty() {
                        "-".to_string()
                    } else {
                        s.shape
                            .iter()
                            .map(|d| d.to_string())
                            .collect::<Vec<_>>()
                            .join(",")
                    };
                    out.push_str(&format!(
                        "{} {} {} {}\n",
                        s.data_offset, s.data_len, s.dtype, shape
                    ));
                }
                std::io::stdout()
                    .write_all(out.as_bytes())
                    .expect("failed to write stdout");
            }
            Err(e) => {
                eprintln!("tensor-spans decode error: {}", e);
                std::process::exit(1);
            }
        }
        return;
    }

    // --- Cowrie FILE container modes (SPEC-v1 §7) ---
    if file_id || file_recode {
        match decode_file(&data, true) {
            Ok(frames) => {
                if file_id {
                    // Print the verified Merkle root as one line of lowercase hex (68 chars).
                    println!("{}", to_hex(&merkle_root(&frames)));
                } else {
                    // Re-encode the file; canonical input reproduces byte-for-byte.
                    let out = encode_file(&frames);
                    std::io::stdout().write_all(&out).expect("failed to write stdout");
                }
            }
            Err(e) => {
                eprintln!("{}: file decode error: {}", e.err_code(), e);
                std::process::exit(1);
            }
        }
        return;
    }

    let decoded = if strict {
        decode_framed_strict(&data)
    } else {
        decode_framed(&data)
    };

    match decoded {
        Ok(v) => {
            // --- §4 schema fingerprint mode: print fingerprint64 as 16 lowercase hex chars ---
            if fingerprint_mode {
                println!("{:016x}", schema_fingerprint64(&v));
                return;
            }
            match encode(&v) {
            Ok(out) => {
                if addr {
                    // Content address of the canonical bytes we would otherwise emit.
                    let address = address_of_bytes(&out);
                    println!("{}", to_hex(&address));
                } else {
                    std::io::stdout().write_all(&out).expect("failed to write stdout");
                }
            }
            Err(e) => {
                eprintln!("{}: encode error: {}", e.err_code(), e);
                std::process::exit(1);
            }
            }
        }
        Err(e) => {
            eprintln!("{}: {}", e.err_code(), e);
            std::process::exit(1);
        }
    }
}

File: tools/cowrie_ref/recode.py
"""``recode`` CLI — the Python v1 (COWR) command, a peer to the Go/Rust/TS ``recode`` tools.

Reads wire/file bytes on **stdin** and supports the EXACT same contracts the other languages'
``recode`` CLIs use (see ``rust/examples/recode.rs`` and ``go/cmd/cowrie`` ``recodeCmd``). It is a
thin shell over the existing :mod:`cowrie_ref` primitives — it NEVER reimplements the codec:

  bare ``recode``      decode (lenient §5.3) -> re-encode canonical bytes to stdout
  ``--strict``         strict decode (§5.3): reject non-canonical, print ``ERR_*`` to stderr, exit 1
  ``--addr``           content address (§3): multihash SHA-256 hex of the canonical bytes
  ``--fingerprint``    schema fingerprint64 (§4) as 16 lowercase hex chars
  ``--tensor-spans``   Phase-2 zero-copy tensor locator: one line per tensor, document order,
                       ``<data_offset> <data_len> <dtype> <shape>`` (shape ``-`` for rank-0)
  ``--file-id``        Cowrie FILE (§7): decode+verify, print merkle_root as lowercase hex
  ``--file-recode``    Cowrie FILE (§7): decode -> re-encode -> raw bytes to stdout
  ``--dataset-root``   read ``uvarint(n) || n*(uvarint(len) || blob)``; print §7 dataset_root hex

``--strict`` may combine with ``--addr``/``--fingerprint`` (the underlying decode is then strict).
``STRICT=1`` in the environment also turns on strict mode (parity with the cross-language harness).
"""
from __future__ import annotations

import os
import sys

from . import (
    address_of_bytes,
    content_address,  # noqa: F401  (re-exported convenience; address_of_bytes is what we use)
    decode,
    encode,
    fingerprint,
    tensor_spans,
)
from . import file as _file
from . import profiles as _profiles
from .errors import CowrieError
from .varint import decode_uvarint


def _read_stdin() -> bytes:
    return sys.stdin.buffer.read()


def _write_stdout(data: bytes) -> None:
    sys.stdout.buffer.write(data)
    sys.stdout.buffer.flush()


def _fail(err: CowrieError) -> int:
    """Print the canonical SPEC-v1 §2.6 ``ERR_*`` code (+ detail) to stderr; the cross-language gate
    extracts the first ``/ERR_[A-Z_]+/`` token from stderr, so we lead with the code (mirrors the
    Go/Rust ``recode`` tools)."""
    sys.stderr.write(f"{err.code}: {err}\n")
    return 1


def _shape_str(shape: tuple[int, ...]) -> str:
    return ",".join(str(d) for d in shape) if shape else "-"


def main(argv: list[str]) -> int:
    flags = set(argv)
    strict = "--strict" in flags or os.environ.get("STRICT") in ("1", "true")

    # --- §7 dataset/stream-layer Merkle root: uvarint(n) || n*(uvarint(len) || blob) ---
    if "--dataset-root" in flags:
        data = _read_stdin()
        pos = 0
        n, pos = decode_uvarint(data, pos)
        blobs: list[bytes] = []
        for _ in range(n):
            ln, pos = decode_uvarint(data, pos)
            blobs.append(data[pos : pos + ln])
            pos += ln
        # Reuse the §7 Merkle construction (profiles.dataset_root) — do not reimplement.
        print(_profiles.dataset_root(blobs).hex())
        return 0

    # --- Phase-2 zero-copy tensor locator: one line per tensor, document order ---
    if "--tensor-spans" in flags:
        data = _read_stdin()
        try:
            spans = tensor_spans(data, strict=strict)
        except CowrieError as e:
            return _fail(e)
        out = "".join(
            f"{data_off} {data_len} {dtype} {_shape_str(shape)}\n"
            for (dtype, shape, data_off, data_len) in spans
        )
        _write_stdout(out.encode("ascii"))
        return 0

    # --- Cowrie FILE container modes (§7) ---
    if "--file-id" in flags or "--file-recode" in flags:
        data = _read_stdin()
        try:
            frames = _file.decode_file(data, verify=True)
            if "--file-id" in flags:
                print(_file.merkle_root(frames).hex())
            else:
                # Re-encode the file; canonical input reproduces byte-for-byte. Each frame is a
                # complete canonical §2.2 value, so decode it back to a value then re-encode_file.
                values = [decode(f) for f in frames]
                _write_stdout(_file.encode_file(values))
        except CowrieError as e:
            return _fail(e)
        return 0

    # --- value modes: decode (lenient or strict) then act ---
    data = _read_stdin()
    try:
        value = decode(data, strict=strict)
    except CowrieError as e:
        return _fail(e)

    if "--fingerprint" in flags:
        h64, _ = fingerprint(value)
        print(f"{h64:016x}")
        return 0

    canonical = encode(value)

    if "--addr" in flags:
        print(address_of_bytes(canonical).hex())
        return 0

    _write_stdout(canonical)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
